apps/dcos_net/src/dcos_net_mesos_listener.erl

```diff
@@ -586,30 +600,33 @@ merge_task_ports(PortMappings, PortResources, DiscoveryPorts) ->
 merge_task_ports([], DiscoveryPorts) ->
     DiscoveryPorts;
 merge_task_ports(PortMappings, DiscoveryPorts) ->
-    Ports =
-        maps:from_list([ begin
-            A = maps:get(protocol, TaskPort),
-            B = maps:get(port, TaskPort, undefined),
-            C = maps:get(host_port, TaskPort, undefined),
-            {{A, B, C}, TaskPort}
-        end || TaskPort <- DiscoveryPorts ]),
-    Ports0 =
-        lists:foldl(fun (TaskPort, Acc) ->
-            A = maps:get(protocol, TaskPort),
-            B = maps:get(port, TaskPort, undefined),
-            C = maps:get(host_port, TaskPort, undefined),
-            KeyA = {A, B, undefined},
-            KeyB = {A, undefined, C},
-            KeyC = {A, B, C},
-            case mfind([KeyA, KeyB, KeyC], Acc) of
-                {ok, Key, TP} ->
-                    TP0 = merge_ports(TP, TaskPort),
-                    maps:put(Key, TP0, Acc);
-                error ->
-                    maps:put(KeyC, TaskPort, Acc)
-            end
-        end, Ports, PortMappings),
-    maps:values(Ports0).
+    lists:foldl(fun (Port, Acc) ->
+        merge_task_port(Port, Acc, [])
+    end, DiscoveryPorts, PortMappings).
+
+-spec(merge_task_port(P, [P], [P]) -> [P] when P :: task_port()).
+merge_task_port(Port, [], Acc) ->
+    [Port|Acc];
+merge_task_port(#{protocol := _P} = PortA, [PortB|Ports], Acc) ->
+    case match_task_port(PortA, PortB) of
+        true ->
+            PortC = merge_ports(PortB, PortA),
+            merge_task_port(PortC, [], Acc ++ Ports);
+        false ->
+            merge_task_port(PortA, Ports, [PortB|Acc])
+    end;
+merge_task_port(Port, Ports, Acc) ->
```

Comment:
any reason why we are not ignoring the ports without `protocol`?

Reply:
That's a good question, we can ignore them. Let me update the PR.